Add support for all Unicode codepoints of the BMP in the JSON parser [TG-7634]

antlechner opened the pull request:
This is a follow-up PR to https://github.com/diffblue/cbmc/pull/4590, which implemented support in the JSON parser for hexadecimal Unicode representations (`\uABCD`) but was limited to codepoints <= `0x7f` (with values from `0x80` to `0xff` also sort of working, representing them as integers rather than characters).
This improved implementation supports all values from `\u0000` to `\uFFFF`, i.e. all codepoints in the BMP.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

antlechner committed:
Add support for whole BMP in the JSON parser

The previous implementation only supported codepoints up to 0x7f as
characters, and all remaining codepoints up to 0xff as integers.
The new implementation supports all codepoints in the BMP, i.e. up to
0xffff.

tautschnig commented on src/util/unicode.h:
```diff
@@ -31,6 +31,11 @@ std::string utf16_native_endian_to_java(const std::wstring &in);
 std::vector<std::string> narrow_argv(int argc, const wchar_t **argv_wide);
 
+/// \param hex: a representation of a UTF-16 character as a four-digit string
+///   (e.g.\ "0041" for \u0041)
```
It seems you will need to escape this for Doxygen to be happy.

tautschnig commented on src/util/unicode.cpp:
```diff
@@ -315,3 +318,12 @@ std::string utf16_native_endian_to_java(const std::wstring &in)
+  PRECONDITION(hex.length() == 4);
+  const char16_t utf16_char = std::strtol(hex.c_str(), nullptr, 16);
+  const std::u16string u16(1, utf16_char);
+  return std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t>{}
```
This is failing with a linker error with Visual Studio. It's a known problem and https://stackoverflow.com/questions/32055357/visual-studio-c-2015-stdcodecvt-with-char16-t-or-char32-t discusses some fixes. It seems that using `wchar_t` instead of `char16_t` might be the easiest of those, if that does actually work. But it would need to be wrapped in an `#ifdef _MSC_VER`.

antlechner replied:
I tried the solution that uses `wchar_t`, the AWS Windows build is still failing and I can't tell from the logs if it's because of this or for some other reason... Can you tell from the logs what the problem is?

antlechner replied:
Turns out the problem was a compilation error that, for some reason, did not show up anywhere in the AWS build logs.
Now everything builds fine but test are failing - again only on Windows...

antlechner replied:
The problem was the encoding of string literals, which can vary between compilers. Should all be fixed now!

romainbrenguier commented on unit/json/json_parser.cpp:
```diff
@@ -84,17 +84,21 @@ SCENARIO("Loading JSON files")
     }
   }
-  GIVEN("A JSON file containing a hexadecimal Unicode character")
+  GIVEN("A JSON file containing hexadecimal Unicode characters")
```
:pick: The use of "characters" here is a bit confusing. I would talk about symbols (or code points) for unicode and only use characters for the c++ `char` type :hamburger:

romainbrenguier commented on src/util/unicode.h:
```diff
@@ -31,6 +31,11 @@ std::string utf16_native_endian_to_java(const std::wstring &in);
 
 std::vector<std::string> narrow_argv(int argc, const wchar_t **argv_wide);
 
+/// \param hex: a representation of a UTF-16 character as a four-digit string
```
:hamburger:

romainbrenguier commented on src/util/unicode.h:
```diff
@@ -31,6 +31,11 @@ std::string utf16_native_endian_to_java(const std::wstring &in);
 
+/// \param hex: a representation of a UTF-16 character as a four-digit string
+///   (e.g.\ "0041" for \u0041)
+/// \return the UTF-8 encoding of the character.
```
:hamburger:

smowton commented:
Nitpick: chars between 00 and ff are not UTF-8 (which can represent all unicode characters, perhaps with many bytes) -- better to just say codepoints <= 255 or 0xff. Similarly codepoints <= 0xffff are the basic multilingual plane (BMP), but not UTF-16, which again can represent all codepoints, perhaps using multiple wchars. Therefore I suggest twiddling PR description / commit messages to clarify that this is adding support for all BMP unicode codepoints

smowton commented on src/json/parser.y:
```diff
@@ -51,7 +53,7 @@ static std::string convert_TOK_STRING()
         assert(p + 4 < yyjsontext + len - 1);
         std::string hex(++p, 4);
-        result += std::stoi(hex, nullptr, 16);
+        result += utf16_hex_to_utf8(hex);
```
Rename codepoint_hex_to_utf8 -- the difference is that utf-16 implies the possibility of two 16-bit numbers representing a single codepoint, such as U+10437 == 0xD801 0xDC37

smowton commented on src/util/unicode.cpp:
```diff
@@ -315,3 +318,25 @@ std::string utf16_native_endian_to_java(const std::wstring &in)
   return result.str();
 }
+
+char16_t utf16_hex_to_utf16_native_endian(const std::string &hex)
```
Again, codepoint rather than utf16

smowton commented on src/util/unicode.h:
```diff
@@ -31,6 +31,17 @@ std::string utf16_native_endian_to_java(const std::wstring &in);
+
+std::string utf16_native_endian_to_utf8(char16_t utf16_char);
+
+/// \param hex: a representation of a UTF-16 symbol as a four-digit string
```
of a BMP codepoint

antlechner committed:
Name last_hex_digit variable in JSON parser

A named variable is clearer that referring to the value twice in
relation to p, where p changes in between.

antlechner commented:
All review comments have been addressed. I split the original "hex to UTF8" function into four tiny functions, as this makes it easier to reuse them (and I'm planning to use one of them in some new code soon). All four functions are documented (@smowton could you check that the terminology in the Doxygen makes sense?). Some of the string literals in the unit test have been specified to use UTF-8 encoding as the Windows compiler used a different encoding by default.
I also addressed @smowton 's review comment from https://github.com/diffblue/cbmc/pull/4590, I used a variable name instead of comments to (hopefully) clarify how the parser processes Unicode representations.

smowton reviewed (approved):
I'd probably clarify that while anything taking a `char16_t` *is* converting UTF-16 to UTF-8, it's not tackling the whole of UTF-16, which can use low / high surrogate pairs (two successive char16_ts) to represent characters outside the BMP (i.e. those from U+0x10000 upwards). So really they're converting a UTF-16 BMP char... which is always just equal to the codepoint number. Otherwise lgtm.

smowton commented:
(In case that was a bit confused... what I'm saying is we're handling that subset of Unicode where the UTF-16 <-> codepoint number transformation is trivial (`int codepoint = 0x1234; char16_t utf16_codepoint = (char16_t)codepoint;`). By contrast those chars that don't fit into a char16_t need a more involved transformation.
